pkg/services/sqlstore/migrations/ualert/channel.go

```diff
@@ -0,0 +1,261 @@
+package ualert
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+
+	"github.com/prometheus/alertmanager/pkg/labels"
+
+	"github.com/grafana/grafana/pkg/components/securejsondata"
+	"github.com/grafana/grafana/pkg/components/simplejson"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util"
+)
+
+type notificationChannel struct {
+	Uid                   string                        `xorm:"uid"`
+	Name                  string                        `xorm:"name"`
+	Type                  string                        `xorm:"type"`
+	DisableResolveMessage bool                          `xorm:"disable_resolve_message"`
+	IsDefault             bool                          `xorm:"is_default"`
+	Settings              *simplejson.Json              `xorm:"settings"`
+	SecureSettings        securejsondata.SecureJsonData `xorm:"secure_settings"`
+}
+
+func (m *migration) getNotificationChannelMap() (map[string]*notificationChannel, *notificationChannel, error) {
+	q := `
+	SELECT uid,
+		name,
+		type,
+		disable_resolve_message,
+		is_default,
+		settings,
+		secure_settings
+	FROM
+		alert_notification
+	`
+	allChannels := []notificationChannel{}
+	err := m.sess.SQL(q).Find(&allChannels)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if len(allChannels) == 0 {
+		return nil, nil, nil
+	}
+
+	allChannelsMap := make(map[string]*notificationChannel)
+	var defaultChannel *notificationChannel
+	for i, c := range allChannels {
+		allChannelsMap[c.Uid] = &allChannels[i]
+		if c.IsDefault {
+			// TODO: verify that there will be only 1 default channel.
+			defaultChannel = &allChannels[i]
+		}
+	}
+
+	if defaultChannel == nil {
+		// TODO: is this possible?
+		defaultChannel = &allChannels[0]
+	}
+
+	return allChannelsMap, defaultChannel, nil
+}
+
+func (m *migration) makeReceiverAndRoute(ruleUid string, channelUids []string, allChannels map[string]*notificationChannel) (*PostableApiReceiver, *Route, error) {
+	receiverName := getMigratedReceiverNameFromRuleUID(ruleUid)
+
+	portedChannels := []*PostableGrafanaReceiver{}
+	receiver := &PostableApiReceiver{
+		Name: receiverName,
+	}
+
+	for _, n := range channelUids {
+		m, ok := allChannels[n]
+		if !ok {
+			// TODO: should we error out here?
```

Comment:
No, I think it shouldn't fail here: references to obsolete notifiers inside dashboard JSON are quite common.